### DISCUSSION TITLE
[Q] How does the concurrency aspect actually work?

### DISCUSSION THREAD

**Discussion by fabis94** [roadrunner-server/roadrunner#788] [Q&A] [answered]
RoadRunner advertises high performance through multi-threading and coroutines, but how does this actually work at the PHP level? I couldn't find any resources that explain this in detail, but if there are any I'd love to read them.

Basically, how does RoadRunner ensure that a single PHP app can actually work on multiple requests at the same time, instead of being blocked by a single process at a time like classic PHP?

**Comment by rustatian** [comment, marked as answer]
A simple diagram is here: https://github.com/spiral/roadrunner/discussions/622#discussioncomment-561756

RR runs a set of processes with PHP workers inside. The workers are similar and path to them specified by the server: https://github.com/spiral/roadrunner-binary/blob/master/.rr.yaml#L23. So, there is no single PHP app, there are a set of processes with a PHP inside and RR manages these processes and extends functionality by the plugins.

You can find simplified docs here: https://github.com/spiral/roadrunner/tree/master/pkg/doc

**Reply by fabis94** [reply]
I see, so a single PHP worker runs a while loop that retrieves incoming requests from the Go side. And the Go side scales worker scripts up/down depending on config and the amount of incoming requests. Thanks!

**Reply by rustatian** [reply]
Yep, with the exception, that at the moment we do not scale (up or down) a number of the workers. They are relatively cheap (costs only memory). 
EDIT:
The number of workers is static per RR run.

**Reply by Warxcell** [reply]
Costs only memory till they are in idle, no? So if it happens to have 100 workers and 1 CPU - and we have CPU intensive code and at same time all workers receive request to process - it will slow down the whole system.

**Reply by rustatian** [reply]
> So if it happens to have 100 workers and 1 CPU - and we have CPU intensive code and at same time all workers receive request to process - it will slow down the whole system.

Every process in the system slows down the whole system. It's a bad choice to use 100 workers when you have only 1 CPU (with CPU intensive tasks), then all tasks will be processed concurrently, not in parallel.

**Reply by Warxcell** [reply]
The problem we have is that we have CPU bound tasks AND IO bound tasks. So you cannot either increase the number of workers - best bet would be to scale horizontally depending on free workers. But RoadRunner doesn't expose that statistics via metrics plugin.

**Reply by rustatian** [reply]
Yeah, in that case, will be better to use different RR instances in different pods. One pod may receive a computing-related task with a big number of CPUs, another, less number of CPU, but much more memory to handle IO related tasks with a bigger number of workers (they won't consume a lot of CPU, so, it's safe to start 2-3-n+1 RR instances on the same node with an identical config (vertical scaling, `SO_REUSEPORT` supported by the RR)). So, it's just a simple DevOps task 😃

**Reply by Warxcell** [reply]
Unless if you have CPU and IO bound task under same url. I would be very happy if RoadRunner expose statistics of free/busy workers via metrics plugin. It will allow to bring up instances depending on how many workers are busy.

**Reply by rustatian** [reply]
Feature requests are welcome 😃

**Reply by rustatian** [reply]
@Warxcell So, I guess, that you want smt like that:
```log
# HELP rr_http_worker_status_25727 Worker current state
# TYPE rr_http_worker_status_25727 gauge
rr_http_worker_status_25727{status="ready"} 0
# HELP rr_http_worker_status_25731 Worker current state
# TYPE rr_http_worker_status_25731 gauge
rr_http_worker_status_25731{status="ready"} 0
# HELP rr_http_worker_status_25732 Worker current state
# TYPE rr_http_worker_status_25732 gauge
rr_http_worker_status_25732{status="ready"} 0
# HELP rr_http_worker_status_25734 Worker current state
# TYPE rr_http_worker_status_25734 gauge
rr_http_worker_status_25734{status="ready"} 0
# HELP rr_http_worker_status_25737 Worker current state
# TYPE rr_http_worker_status_25737 gauge
rr_http_worker_status_25737{status="ready"} 0
# HELP rr_http_worker_status_25738 Worker current state
# TYPE rr_http_worker_status_25738 gauge
rr_http_worker_status_25738{status="ready"} 0
# HELP rr_http_worker_status_25739 Worker current state
# TYPE rr_http_worker_status_25739 gauge
rr_http_worker_status_25739{status="ready"} 0
# HELP rr_http_worker_status_25740 Worker current state
# TYPE rr_http_worker_status_25740 gauge
rr_http_worker_status_25740{status="ready"} 0
# HELP rr_http_worker_status_25741 Worker current state
# TYPE rr_http_worker_status_25741 gauge
rr_http_worker_status_25741{status="ready"} 0
# HELP rr_http_worker_status_25742 Worker current state
# TYPE rr_http_worker_status_25742 gauge
rr_http_worker_status_25742{status="ready"} 0
# HELP rr_http_workers_memory_bytes Memory usage by HTTP workers.
# TYPE rr_http_workers_memory_bytes gauge
rr_http_workers_memory_bytes 2.65252864e+08
```
Where `r_http_worker_status_25742` - current worker state (number at the end is PID).
And `{status="ready"}` is the current worker's state. (working, invalid, killing, destroyed, ready, etc).

**Reply by rustatian** [reply]
Or even better:
```log
rr_http_worker_memory_bytes{pid="47164"} 2.6013696e+07
rr_http_worker_memory_bytes{pid="47166"} 2.6386432e+07
rr_http_worker_memory_bytes{pid="47168"} 2.7410432e+07
rr_http_worker_memory_bytes{pid="47172"} 2.6349568e+07
rr_http_worker_memory_bytes{pid="47173"} 2.6783744e+07
rr_http_worker_memory_bytes{pid="47174"} 2.7267072e+07
rr_http_worker_memory_bytes{pid="47175"} 2.7066368e+07
rr_http_worker_memory_bytes{pid="47176"} 2.654208e+07
rr_http_worker_memory_bytes{pid="47177"} 2.6914816e+07
rr_http_worker_memory_bytes{pid="47178"} 2.656256e+07
# HELP rr_http_worker_state Worker current state
# TYPE rr_http_worker_state gauge
rr_http_worker_state{pid="47164",state="ready"} 0
rr_http_worker_state{pid="47166",state="ready"} 0
rr_http_worker_state{pid="47168",state="ready"} 0
rr_http_worker_state{pid="47172",state="ready"} 0
rr_http_worker_state{pid="47173",state="ready"} 0
rr_http_worker_state{pid="47174",state="ready"} 0
rr_http_worker_state{pid="47175",state="ready"} 0
rr_http_worker_state{pid="47176",state="ready"} 0
rr_http_worker_state{pid="47177",state="ready"} 0
rr_http_worker_state{pid="47178",state="ready"} 0
```

**Reply by Warxcell** [reply]
that would be great, yes. this is how PHPFPM exporter does it:

phpfpm_total_processes{pool="www",scrape_uri="tcp://127.0.0.1:9000/status"} 5
phpfpm_active_processes{pool="www",scrape_uri="tcp://127.0.0.1:9000/status"} 1
phpfpm_idle_processes{pool="www",scrape_uri="tcp://127.0.0.1:9000/status"} 4

- but your variant should be able to aggregate same values.

**Reply by rustatian** [reply]
Yeah, PHP-FPM has only 3 states, while RR worker has much more. You may track for example workers which are in the stopped state and are prepared to reallocate.
I will add total workers as well.

**Reply by rustatian** [reply]
So, ok, the final stats will be as following:
```log
# HELP rr_http_total_workers Total number of workers used by the HTTP plugin
# TYPE rr_http_total_workers gauge
rr_http_total_workers 10
# HELP rr_http_worker_memory_bytes Worker current memory usage
# TYPE rr_http_worker_memory_bytes gauge
rr_http_worker_memory_bytes{pid="58137"} 2.7758592e+07
rr_http_worker_memory_bytes{pid="58140"} 2.5870336e+07
rr_http_worker_memory_bytes{pid="58143"} 2.6062848e+07
rr_http_worker_memory_bytes{pid="58146"} 2.648064e+07
rr_http_worker_memory_bytes{pid="58147"} 2.711552e+07
rr_http_worker_memory_bytes{pid="58148"} 2.725888e+07
rr_http_worker_memory_bytes{pid="58149"} 2.67264e+07
rr_http_worker_memory_bytes{pid="58150"} 2.7168768e+07
rr_http_worker_memory_bytes{pid="58151"} 2.7148288e+07
rr_http_worker_memory_bytes{pid="58152"} 2.6959872e+07
# HELP rr_http_worker_state Worker current state
# TYPE rr_http_worker_state gauge
rr_http_worker_state{pid="58137",state="ready"} 0
rr_http_worker_state{pid="58140",state="ready"} 0
rr_http_worker_state{pid="58143",state="ready"} 0
rr_http_worker_state{pid="58146",state="ready"} 0
rr_http_worker_state{pid="58147",state="ready"} 0
rr_http_worker_state{pid="58148",state="ready"} 0
rr_http_worker_state{pid="58149",state="ready"} 0
rr_http_worker_state{pid="58150",state="ready"} 0
rr_http_worker_state{pid="58151",state="ready"} 0
rr_http_worker_state{pid="58152",state="ready"} 0
```
